src/openms/source/ANALYSIS/MAPMATCHING/MapAlignmentAlgorithmTreeGuided.cpp

```diff
@@ -0,0 +1,291 @@
+  class MapAlignmentAlgorithmTreeGuided::PeptideIdentificationsPearsonDistance_
+  {
+  public:
+    float operator()(SeqAndRTList& map_first, SeqAndRTList& map_second) const
+    {
+      // create vectors for both maps containing RTs of identical peptide sequences and
+      // get union and intercept amount of peptides
+      auto pep1_it = map_first.begin();
+      auto pep2_it = map_second.begin();
+      vector<double> intercept_rts1;
+      vector<double> intercept_rts2;
+      float union_size = 0.0;
+      while (pep1_it != map_first.end() && pep2_it != map_second.end())
+      {
+        if (pep1_it->first < pep2_it->first)
+        {
+          ++pep1_it;
+        }
+        else if (pep2_it->first < pep1_it->first)
+        {
+          ++pep2_it;
+        }
+        else
+        {
+          double med1 = Math::median(pep1_it->second.begin(), pep1_it->second.end(), true);
+          intercept_rts1.push_back(med1);
+          double med2 = Math::median(pep2_it->second.begin(), pep2_it->second.end(), true);
+          intercept_rts2.push_back(med2);
+          ++pep1_it;
+          ++pep2_it;
+        }
+        ++union_size;
+      }
+      Size intercept_size = intercept_rts1.size();
+
+      // pearsonCorrelationCoefficient(rt_map_i, rt_map_j)
+      float pearson_val;
+      pearson_val = static_cast<float>(Math::pearsonCorrelationCoefficient(intercept_rts1.begin(), intercept_rts1.end(),
+                                                                     intercept_rts2.begin(), intercept_rts2.end()));
+      if (pearson_val > 1)
+      {
+        throw Exception::InvalidRange(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION);
+      }
+      // Small intersections are penalized by multiplication with the quotient of intersection to union.
+      return pearson_val * intercept_size / union_size;
+    }
+  }; // end of PeptideIdentificationsPearsonDifference
+
+  // For given peptide identifications extract sequences and store with associated feature RT.
+  void MapAlignmentAlgorithmTreeGuided::addPeptideSequences_(const std::vector<PeptideIdentification> &peptides,
+          SeqAndRTList &peptide_rts, std::vector<double> &map_range, double feature_rt)
+  {
+    for (const auto & peptide : peptides)
+    {
+      if (!peptide.getHits().empty())
+      {
+        const String& sequence = peptide.getHits()[0].getSequence().toString();
+        peptide_rts[sequence].push_back(feature_rt);
+        map_range.push_back(feature_rt);
+      }
+    }
+  }
+
+  // For each input map, extract peptide identifications (sequences) of existing features with associated feature RT.
+  void MapAlignmentAlgorithmTreeGuided::extractSeqAndRt_(const vector<FeatureMap>& feature_maps,
+          vector<SeqAndRTList>& maps_seq_and_rt, vector<vector<double>>& maps_ranges)
+  {
+    for (Size i = 0; i < feature_maps.size(); ++i)
+    {
+      for (auto feature_it = feature_maps[i].begin(); feature_maps[i].end() != feature_it; ++feature_it)
```

Comment:
should be possible to write loop without iterators here for(const FeatureMap& fm : feature_maps)